### PR TITLE
Unit tests for Job app

### PR DESCRIPTION
**Pull request by MAM-SYS**
hello friends
i wrote some unit tests for Views and Models of Job application for #12 
hope it can help
I can also change anything you think is needed in these tests to get better
:v:

### CODE REVIEW

**Comment by sbabashahi** [comment]
Hello dear @MAM-SYS  

sorry for late response, I will check your MR asap.

**Comment by MAM-SYS** [comment]
thank you @sbabashahi 
Let me know if there is anything needs to be fix in UnitTests
i will make it better

File: job/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            "name": "برنامه نویس پایتون",
+            "cover": "/default/system/cover.jpg",
+            "icon": "fas fa-home",
+            "description": "برنامه نویس پایتون",
+        }
+        self.client.credentials(HTTP_AUTHORIZATION="JWT " + self.token)
+        response = self.client.post(url, data, format="json")
+        job = Job.objects.last()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response._headers["content-type"][1], "application/json")
```
Comment (sbabashahi) [inline job/tests/test_views.py]: simply response['content-type']

File: job/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        self.assertEqual(Job.objects.count(), 3)
+        self.assertEqual(job.name, "برنامه نویس پایتون")
+        self.assertEqual(job.job_slug, slug_helper("برنامه نویس پایتون"))
+
+    def test_job_listview(self):
```
Comment (sbabashahi) [inline job/tests/test_views.py]: Suggestion: to have integrity in names try use a format like test_instance_functionality_view_other_description: test_job_list_view

File: job/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        # No params passed to the request
+        url = "/job/list/"
+        self.client.credentials(HTTP_AUTHORIZATION="JWT " + self.token)
+        response = self.client.get(url)
+        response_data = response.json()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertContains(response, "is_deleted", 2)
```
Comment (sbabashahi) [inline job/tests/test_views.py]: something like 
``       self.assertEqual(list(response.json()['data'][0].keys()),
                         ['id', 'name', 'job_slug', 'cover', 'icon', 'description',
                          'is_deleted', 'approved'])`` could check for fields we need in response

File: job/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        url = f"/job/list/?index={index}"
+        response = self.client.get(url)
+        response_data = response.json()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response_data["data"], [])
+        self.assertEqual(response_data["index"], 100)
+        self.assertEqual(response_data["total"], 2)
+        
+        
+
```
Comment (sbabashahi) [inline job/tests/test_views.py]: try pep8, remove redundant lines

File: job/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        self.assertEquals(response_data["data"], data)
+
+        ## Updating non-existent object
+        url = "/job/5000/update/"
+        response = self.client.patch(url, partial_data, format="json")
+        response_data = response.json()
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        self.assertEqual(response_data["message"], "Instance does not exist.")
+
+    def test_job_delete(self):
```
Comment (sbabashahi) [inline job/tests/test_views.py]: you could divide this test and others to multi tests for example test_job_delete_view_when_exists & test_job_delete_view_when_not_exists try this on other tests too.

**Comment by sbabashahi** [comment]
First of all thanks for your contribution :) and sorry again for late response.
1. These tests are not unit tests, they are integration and functional tests, it's not a big deal, we appreciate writing any type of test.
2. it could be a good practice to create a special DJANGO_SETTINGS_MODULE for testing (because you are the first one writing test and it could help others) a quick hint:
 write a tests.py in settings and in manage.py something like:

`if 'test' in sys.argv:
  os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'ratecompany.settings.tests')
else:
  os.environ.setdefault('DJANGO_SETTINGS_MODULE',
                        'ratecompany.settings.development')`
3. Hint for unit test, test for default, required and optional fields.
As I said we appreciate your work and we could merge it right when you resolve conversations, but if you could improve it (add test settings and more test) we could wait. so your call :)

**Comment by MAM-SYS** [comment]
@sbabashahi 
thank you boss.thanks for your guidance
certainly i will add and change them as you suggested to be better
it's my pleasure

**Comment by MAM-SYS** [comment]
Hello friends @sbabashahi 
i fixed the codes as you mentioned
hope it is better now
but if still anything else that needs to be fix or add; just say
have a good Friday

File: ratecompany/settings/tests.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+SECRET_KEY = "some-khafan-key"
+
+EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"
+
+SMS_BACKEND = "CONSOLE"
+
+DATABASES = {
+    "default": {
+        "ENGINE": "django.contrib.gis.db.backends.postgis",
```
Comment (sbabashahi) [inline ratecompany/settings/tests.py:16]: Why you changed name and password?

File: job/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        }
+        self.client.credentials(HTTP_AUTHORIZATION="JWT " + self.token)
+        url = "/job/5000/update/"
+        response = self.client.patch(url, partial_data, format="json")
+        response_data = response.json()
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        self.assertEqual(response_data["message"], "Instance does not exist.")
+
+    def test_job_delete_view_when_object_exists(self):
+        # Deleting an existent object
```
Comment (sbabashahi) [inline job/tests/test_views.py]: I think these comments are redundant now, you explained your test functionality on test name

File: ratecompany/settings/tests.py (hunk 6 above)
Comment (MAM-SYS) [reply ratecompany/settings/tests.py:16]: Excuse Me For That
first time that i was trying to work with docker-compose i used the the name and password(in django setting file) that is similar with environments in docker-compose file
after that i forgot to change them again for this PR.i'll make it same as others
sorry
https://github.com/jobguywork/backend/blob/2e8cb704dcabe61d36163bb336b938fa4d05a5b1/docker-compose.yml#L9-L12

File: job/tests/test_views.py (hunk 7 above)
Comment (MAM-SYS) [reply job/tests/test_views.py]: exactly.they will be removed

**Comment by sbabashahi** [comment]
@MAM-SYS Sorry to keep you waiting, I will check it asap.

**Comment by MAM-SYS** [comment]
thanks a lot @sbabashahi
if anything need to change, tell me.even all of the code

**Comment by sbabashahi** [comment]
@MAM-SYS well done :) i approve it in this state. I add some hints that may help us work better
1. Try to set an in cache db for test
2. Separate func test dir from unit test

**Comment by MAM-SYS** [comment]
@sbabashahi  thanks again for your help in this issue
i consider your hints for my other pull requests for tests

**Pull request merged**